### PR TITLE
Apexmalhar 2458 Move kafka related examples from datatorrent examples to apex-malhar examples

### PR DESCRIPTION
**Pull request by prasannapramod**
@tweise @amberarrow @ashwinchandrap please review

### CODE REVIEW

**Comment by tweise** [comment]
@prasannapramod please rebase this PR, update version number. Are the tests adequate and do they pass? If everything is ready please make it ready for merge so that the related PR 591 can proceed.

File: examples/kafka/pom.xml [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    kafka2hdfs is a example Use of New Kafka Input Operator.
+    hdfs2kafka is a simple application to transfer data from HDFS to Kafka
+  </description>
+  
+  <dependencies>
+    <!-- add your dependencies here -->
+    <dependency>
+      <groupId>org.apache.apex</groupId>
+      <artifactId>malhar-kafka</artifactId>
+      <version>3.6.0</version> <!--check with pramod once-->
```
Comment (tweise) [inline examples/kafka/pom.xml]: this should use ${project.version}

File: examples/kafka/pom.xml [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+      <version>${apex.core.version}</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.kafka</groupId>
+      <artifactId>kafka_2.11</artifactId>
+      <version>0.9.0.1</version>
+    </dependency>
+    <!-- Ask pramod about this having 2 versions of kafka -->
+    <dependency>
```
Comment (tweise) [inline examples/kafka/pom.xml]: I would prefer to have the examples based on Kafka 0.9 operators.

**Comment by tweise** [comment]
@prasannapramod would it be possible to take this PR forward before opening more example PRs?

**Comment by prasannapramod** [comment]
@tweise @amberarrow @ashwinchandrap please review fixed issues.

File: examples/kafka/pom.xml [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  <parent>
+    <groupId>org.apache.apex</groupId>
+    <artifactId>malhar-examples</artifactId>
+    <version>3.8.0-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>malhar-examples-kafka</artifactId>
+  <packaging>jar</packaging>
+
+  <!-- change these to the appropriate values -->
```
Comment (tweise) [inline examples/kafka/pom.xml]: remove this comment?

File: examples/kafka/pom.xml (hunk 1 above)
Comment (tweise) [reply examples/kafka/pom.xml]: still to be addressed

File: examples/kafka/pom.xml [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  
+  <dependencies>
+    <!-- add your dependencies here -->
+    <dependency>
+      <groupId>org.apache.apex</groupId>
+      <artifactId>malhar-kafka</artifactId>
+      <version>3.6.0</version> <!--check with pramod once-->
+      <!--version>${project.parent.version}</version-->
+      <exclusions>
+        <exclusion>
```
Comment (tweise) [inline examples/kafka/pom.xml]: why is this necessary?

File: examples/kafka/pom.xml [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    </dependency>
+    <dependency>
+      <groupId>org.apache.apex</groupId>
+      <artifactId>apex-engine</artifactId>
+      <version>${apex.core.version}</version>
+      <scope>test</scope>
+    </dependency>
+    <!-- -->
+    <dependency> 
+      <groupId>org.apache.kafka</groupId>
```
Comment (tweise) [inline examples/kafka/pom.xml]: Why is this redeclared and not coming from malhar-kafka? Please update malhar-kafka if necessary.

**Review by tweise — CHANGES_REQUESTED** [review]

File: examples/kafka/pom.xml (hunk 4 above)
Comment (prasannapramod) [inline examples/kafka/pom.xml]: @tweise I was doing some reading to understand why this was done. Looks like kafka-clients is the new java implementation for the kafka client API and kafka_<scala_version> is the scala client which we have been using in the older operators for some time.

I can only venture a guess, maybe the author of new kafka operator wanted to make sure that they are API compatible with java API and possibly switch completely to java API and hence they included kafka-clients as a dependency in the kafka operator project pom.xml. However, the author may have found that the java clients don't work correctly for all the functionality that we are using and hence, for example, they are using the scala client for the operator tests, as evidenced by the test dependency pointing to the scala client. I believe the original author of this example also is following the same pattern. They excluded the java client and included the scala one in the dependencies.

Comment (prasannapramod) [inline examples/kafka/pom.xml]: An additional data point. I am looking at 0.9 kafka operator dependencies in the kafka operator pom.xml. I see that there is an "optional" dependency to the Scala api along with the regular dependency to the java api. There is a test dependency to scala api as well like I mentioned above for runtime. 

It didn't seem right that both apis are included as dependencies, albeit one is marked optional. I commented out the scala dependency and noticed that while the operators built fine the tests didn't compile. Surmising that the optional dependency to the scala api for compile was added to successfully build the tests.

File: examples/kafka/pom.xml [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    kafka2hdfs is a example Use of New Kafka Input Operator.
+    hdfs2kafka is a simple application to transfer data from HDFS to Kafka
+  </description>
+  
+  <dependencies>
+    <!-- add your dependencies here -->
+    <dependency>
+      <groupId>org.apache.apex</groupId>
+      <artifactId>malhar-kafka</artifactId>
+      <version>3.6.0</version>
```
Comment (tweise) [inline examples/kafka/pom.xml]: change to project.version

File: examples/kafka/pom.xml [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    <artifactId>malhar-examples</artifactId>
+    <version>3.8.0-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>malhar-examples-kafka</artifactId>
+  <packaging>jar</packaging>
+
+  <name>Apache Apex Malhar Kafka examples</name>
+  <description>
+    kafka2hdfs is a example Use of New Kafka Input Operator.
```
Comment (tweise) [inline examples/kafka/pom.xml]: This description could be improved.

File: examples/kafka/README.md [hunk 8]
```diff
@@ -0,0 +1,16 @@
+## Kafka to HDFS example :
+
+This sample application show how to read lines from a Kafka topic using the new (0.9)
+Kafka input operator and write them out to HDFS using rolling files with a bounded size.
+
+The output files start out with a `.tmp` extension and get renamed when they reach the
+size bound.  Additional operators to perform parsing, aggregation or filtering can be
+inserted into this pipeline as needed.
+
+## HDFS to kafka example :
```
Comment (tweise) [inline examples/kafka/README.md]: kafka => "Kafka"

File: examples/kafka/pom.xml [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+  </description>
+  
+  <dependencies>
+    <!-- add your dependencies here -->
+    <dependency>
+      <groupId>org.apache.apex</groupId>
+      <artifactId>malhar-kafka</artifactId>
+      <version>3.6.0</version> 
+      <exclusions>
+        <exclusion>
```
Comment (tweise) [inline examples/kafka/pom.xml]: I don't see the comments that arrived via email on this PR now. But since you have already researched this, can we somehow capture what should be done in malhar-kafka as follow-up in a separate JIRA. It seem confusing that the user would have to make this exclusion when using the dependency.

Comment (tweise) [reply examples/kafka/pom.xml]: It may be helpful to add appropriate comments in the malhar-kafka pom.xml

**Comment by prasannapramod** [comment]
@tweise addressed the comments. please review.

**Comment by tweise** [comment]
LGTM, please rebase the PR.
